### PR TITLE
Introducing Map for error columns and their values

### PR DESCRIPTION
**Pull request by benedeki**
* Relatively big overwrite to use `map` instead of errCol and sequence of raw values
* `ErrorMessageSubmitJustErrorValue` class created to offer the ability to submit errors without source column but with error value
*  Some other changes to improve the code

After the presentation, discussions with others and more thoughts, I came to the idea to combine the `errCol` (which would be better as sequence anyway) and `rawValues` into one column of `map` data type. That resulted in a need to do some more exercise in the supporting types, but I believe to for the better.

Also some general code improvements were added as I stumbled upon stuff during the refactoring.

### CODE REVIEW

File: spark-commons/src/main/scala/za/co/absa/spark/commons/errorhandling/partials/EvaluateViaUdf.scala [hunk 1]
```diff
@@ -32,13 +32,12 @@ trait EvaluateViaUdf[T] {
...
+     errorMessageSubmit.errColsAndValues.column,
      errorMessageSubmit.additionInfo.column
    )
   }
 }
 
 object EvaluateViaUdf {
-  type ErrorMessageFunction[T] = (ErrType, ErrCode, ErrMsg, ErrCol, RawValues, AdditionalInfo) => T //TODO needed?
+  type ErrorMessageFunction[T] = (ErrType, ErrCode, ErrMsg, ErrColsAndValues, AdditionalInfo) => T //TODO needed?
 }
```
Comment (TebaleloS) [inline spark-commons/src/main/scala/za/co/absa/spark/commons/errorhandling/partials/EvaluateViaUdf.scala:43]: Please a new line after line 43

File: spark-commons/src/main/scala/za/co/absa/spark/commons/errorhandling/partials/EvaluateIntoErrorMessage.scala [hunk 2]
```diff
@@ -18,18 +18,30 @@ package za.co.absa.spark.commons.errorhandling.partials
...
+    val errType = "errType"
+    val errCode = "errCode"
+    val errMsg = "errMsg"
+    val errColsAndValues = "errColsAndValues"
+    val additionInfo = "additionInfo"
+    val errCols = "errCols"
+    val errValues = "errValues"
+  }
 
 }
```
Comment (TebaleloS) [inline spark-commons/src/main/scala/za/co/absa/spark/commons/errorhandling/partials/EvaluateIntoErrorMessage.scala:47]: Please add a new line at the end.

File: spark-commons/src/main/scala/za/co/absa/spark/commons/errorhandling/partials/ErrorHandlingCommon.scala
Comment (TebaleloS) [inline spark-commons/src/main/scala/za/co/absa/spark/commons/errorhandling/partials/ErrorHandlingCommon.scala:1]: Please add a newline at the end (after line 58)

File: spark-commons/src/main/scala/za/co/absa/spark/commons/errorhandling/implementations/ErrorMessageSubmitOnMoreColumns.scala [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+  def apply(errType: ErrType, errCode: ErrCode, errMessage: ErrMsg, errColNames: Set[ErrSourceColName], additionalInfo: AdditionalInfo= None): ErrorMessageSubmitOnMoreColumns = {
+    new ErrorMessageSubmitOnMoreColumns(
+      ColumnOrValue.withValue(errType),
+      ColumnOrValue.withValue(errCode),
+      ColumnOrValue.withValue(errMessage),
+      errColNames,
+      ColumnOrValue.withOption(additionalInfo)
+    )
+  }
+}
```
Comment (TebaleloS) [inline spark-commons/src/main/scala/za/co/absa/spark/commons/errorhandling/implementations/ErrorMessageSubmitOnMoreColumns.scala:44]: Add a new line at the end

File: spark-commons/src/main/scala/za/co/absa/spark/commons/errorhandling/implementations/ErrorMessageSubmitOnColumn.scala [hunk 4]
```diff
@@ -16,33 +16,24 @@
...
-      ColumnOrValue.withActualValue(errCode),
-      ColumnOrValue.withActualValue(errMessage),
-      errSourceColName,
+      ColumnOrValue.withValue(errType),
+      ColumnOrValue.withValue(errCode),
+      ColumnOrValue.withValue(errMessage),
+      errColName,
       ColumnOrValue.withOption(additionalInfo)
     )
   }
```
Comment (TebaleloS) [inline spark-commons/src/main/scala/za/co/absa/spark/commons/errorhandling/implementations/ErrorMessageSubmitOnColumn.scala:39]: Please add a new line after line 39

File: spark-commons/src/main/scala/za/co/absa/spark/commons/errorhandling/types/ColumnOrValue.scala [hunk 5]
```diff
@@ -36,51 +37,62 @@ object ColumnOrValue {
...
+  private final case class CoVNull[T](dataType: DataType) extends ColumnOrValue[T] {
+    val column: Column = null_col(dataType)
+
     val isColumn: Boolean = false
     val isValue: Boolean = true
-    val getColumnName: Option[String] = None
+    val columnNames: Set[String] = Set.empty
     val getValue: Option[T] = None
   }
 }
```
Comment (TebaleloS) [inline spark-commons/src/main/scala/za/co/absa/spark/commons/errorhandling/types/ColumnOrValue.scala:98]: Add a new line at the end

File: spark-commons/src/main/scala/za/co/absa/spark/commons/errorhandling/types/types.scala [hunk 6]
```diff
@@ -20,16 +20,17 @@ import org.apache.spark.sql.Column
...
-  type AdditionalInfo = Option[String] // actually a JSON
+  type ErrColsAndValues = Map[String, String]
+  type AdditionalInfo = Option[JsonString] // actually a JSON
   //mapping is missing, should be part of AdditionalInfo, as being very specific
 
+  type ColumnTransformer = String => Column
+
   //This is to ensure some level of type-safety
   final case class ErrorColumn(column: Column) extends AnyVal
 }
```
Comment (TebaleloS) [inline spark-commons/src/main/scala/za/co/absa/spark/commons/errorhandling/types/types.scala:36]: Add a new line at the end

File: spark-commons/src/main/scala/za/co/absa/spark/commons/sql/functions.scala [hunk 7]
```diff
@@ -63,4 +64,9 @@ object functions {
     lit(None.orNull)
   }
 
+  def null_col(dataType: DataType):Column = {
+    lit(None.orNull).cast(dataType)
+  }
+
+
 }
```
Comment (TebaleloS) [inline spark-commons/src/main/scala/za/co/absa/spark/commons/sql/functions.scala:72]: Please add a new line at the end and remove unnecessary new line (either line 70 or 71)

**Review by TebaleloS — COMMENTED** [review]
Everything else look cool to me except for minor stuff about the unnecessary new lines and adding new lines at the end of every add file.

File: spark-commons/src/main/scala/za/co/absa/spark/commons/errorhandling/partials/EvaluateIntoErrorMessage.scala [hunk 8]
```diff
@@ -18,18 +18,30 @@ package za.co.absa.spark.commons.errorhandling.partials
 
 import org.apache.spark.sql.Column
 import org.apache.spark.sql.functions.struct
-import za.co.absa.spark.commons.errorhandling.ErrorMessageSubmit
+import za.co.absa.spark.commons.errorhandling.{ErrorMessage, ErrorMessageSubmit}
```
Comment (miroslavpojer) [inline spark-commons/src/main/scala/za/co/absa/spark/commons/errorhandling/partials/EvaluateIntoErrorMessage.scala:21]:
```suggestion
import za.co.absa.spark.commons.errorhandling.ErrorMessageSubmit
```
Removed import not used in code.

File: spark-commons/src/main/scala/za/co/absa/spark/commons/errorhandling/types/ColumnOrValue.scala [hunk 9]
```diff
@@ -26,7 +27,7 @@ trait ColumnOrValue[T] {
   def column: Column
   def isColumn: Boolean
   def isValue: Boolean
-  def getColumnName: Option[String]
+  def columnNames: Set[String]
```
Comment (miroslavpojer) [inline spark-commons/src/main/scala/za/co/absa/spark/commons/errorhandling/types/ColumnOrValue.scala:30]:
```suggestion
  def getColumnNames: Set[String]
```
Current method name does not fit to pattern.

**Review by miroslavpojer — COMMENTED** [review]
- [ ] pulled
- [ ] build
- [ ] code review

File: spark-commons/src/main/scala/za/co/absa/spark/commons/errorhandling/implementations/ErrorMessageArray.scala [hunk 10]
```diff
@@ -16,22 +16,54 @@
...
+  }
+
+  private def deMap(arrayCol: Column): Column = transform(arrayCol, decomposeMap)
+  private def reMap(arrayCol: Column): Column = transform(arrayCol, recomposeMap)
+
+  private def appendToErrCol(dataFrame: DataFrame, errorColName: String, colToUnion: Column): DataFrame = {
+    dataFrame.withColumn(errorColName, reMap(array_union(deMap(col(errorColName)), colToUnion)))
+  }
+
+  protected def doTheAggregation(dataFrame: DataFrame, errCols: Column*): DataFrame = {
```
Comment (jirifilip) [inline spark-commons/src/main/scala/za/co/absa/spark/commons/errorhandling/implementations/ErrorMessageArray.scala:62]: I was thinking if we should use the word "aggregation" here and also in `ErrorHandling` trait. So I searched the dictionary for the definition of the word and it fits here but when I first started reading the code, I thought that there would be some `df.groupBy(...).agg(...)`, i.e. some squishing of rows and then some calculation 😄 But if I understood correctly, it's more for combining the errors into one column (or actually, I guess it depends on the implementation).

**Review by jirifilip — COMMENTED** [review]
Hello, David, nice code. I attached my comments but don't take them too seriously, I don't have that much experience in Scala 😃

File: spark-commons/src/main/scala/za/co/absa/spark/commons/errorhandling/ErrorMessage.scala [hunk 11]
```diff
@@ -27,30 +24,14 @@ import za.co.absa.spark.commons.errorhandling.types._
...
  * @param errCode - Internal error code
  * @param errMsg - Textual description of the error
- * @param errCol - The name of the column where the error occurred
- * @param rawValues - Sequence of raw values (which are the potential culprits of the error)
- * @param additionInfo - Sequence of Mappings i.e Mapping Table Column -> Equivalent Mapped Dataset column
+ * @param errColsAndValues - The names of the columns where the error occurred and their raw values (which are the
+ *                         potential culprits of the error)
+ * @param additionInfo - any optional additional information in the form of a JSON string
  */
 case class ErrorMessage(
```
[the remark this answers is not in the document]
Comment (benedeki) [reply spark-commons/src/main/scala/za/co/absa/spark/commons/errorhandling/ErrorMessage.scala:31]: Reason why I used type aliases was that I wasn't sure we are settled on the types. This way it is rather easy to change.

Also the `AdditionalInfo` actually is planned to be more then alias eventually (#90) 

I am happy to discuss the other aliases once the thing is settled (aliases should be easy to remove) 👍

**Pull request merged**